waku/node/peer_manager/peer_manager.nim

```diff
@@ -619,30 +614,53 @@ proc getNumStreams*(pm: PeerManager, protocol: string): (int, int) =
   return (numStreamsIn, numStreamsOut)
 
 proc pruneInRelayConns(pm: PeerManager, amount: int) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
+  let (inRelayPeers, _) = pm.connectedPeers(WakuRelayCodec)
   let connsToPrune = min(amount, inRelayPeers.len)
 
   for p in inRelayPeers[0..<connsToPrune]:
     asyncSpawn(pm.switch.disconnect(p))
 
-proc connectToRelayPeers*(pm: PeerManager) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
-  let maxConnections = pm.switch.connManager.inSema.size
-  let totalRelayPeers = inRelayPeers.len + outRelayPeers.len
-  let inPeersTarget = maxConnections - pm.outRelayPeersTarget
+proc manageRelayPeers*(pm: PeerManager) {.async.} =
+  var peersToConnect: seq[RemotePeerInfo]
+  var peersToDisconnect: int
+
+  for shard in pm.wakuMetadata.shards.items:
+    var peers = pm.peerStore.getPeersByShard(uint16(pm.wakuMetadata.clusterId), uint16(shard))
+
+    peers.keepItIf(it.protocols.contains(WakuRelayCodec))
+
+    let connectedInPeers = peers.filterIt(
+      pm.peerStore.isConnected(it.peerId) and it.direction == Inbound)
+    
+    let connectedOutPeers = peers.filterIt(
+      pm.peerStore.isConnected(it.peerId) and it.direction == Outbound)
+
+    let outPeerDiff = pm.outRelayPeersTarget - connectedOutPeers.len
+    let inPeerDiff = pm.inRelayPeersTarget - connectedInPeers.len
```

Comment:
Do I understand correctly that this now becomes the _per-shard_ targets? That makes sense, except that I think these values are deduced from the underlying absolute maximum connections setting in libp2p. In other words, the switch would have to be initialised to accommodate for the connection targets for each shard (plus allowing some service protocol client slots). This may require having to dynamically altering these targets? In any case, unless I'm missing this, a per-shard target would have to reflect in the switch max-connections configuration.

Reply:
Totally forgot to check where these target value come from. I assumed it was config based. I'll verify.

Reply:
I divided the targets by the number of shards. There's many way to config the node wrong but with the default value it should work fine.